### PR TITLE
internal refactoring

### PR DESCRIPTION
**Pull request by franz**
* refactors compilation callbacks
* portability fixes (Android, MSVC)
* LTTNG tracing improved
* valgrind improvements
* some random fixes

EDIT: removed migrations & changes to memory

### CODE REVIEW

File: lib/CL/pocl_util.c [hunk 1]
```diff
@@ -454,39 +515,49 @@ pocl_create_event_sync (cl_event waiting_event, cl_event notifier_event)
...
-  for (i = 0; i < num_events; ++i)
-    {
-      if (wait_list[i] == NULL)
-        return CL_INVALID_EVENT_WAIT_LIST;
-    }
+  return CL_TRUE;
+}
+
+static cl_int
+pocl_create_command_inner (_cl_command_node **cmd,
```
Comment (pjaaskel) [inline lib/CL/pocl_util.c:549]: This function should have a comment describing what it does exactly.

File: lib/CL/pocl_util.c [hunk 2]
```diff
@@ -517,25 +588,431 @@ cl_int pocl_create_command (_cl_command_node **cmd,
...
+      pocl_create_event_sync ((*event), wle);
+    }
+  POCL_MSG_PRINT_EVENTS (
+      "Created command struct: CMD %p (event %" PRIu64 " / %p, type: %s)\n", *cmd,
+      (*event)->id, *event, pocl_command_to_str (command_type));
+  return CL_SUCCESS;
+}
+
+static int
+pocl_create_migration_commands (cl_device_id dev, cl_event final_event,
```
Comment (pjaaskel) [inline lib/CL/pocl_util.c:607]: ...same here.

File: lib/CL/pocl_cl.h [hunk 3]
```diff
@@ -164,16 +164,27 @@ typedef struct pocl_kernel_metadata_s pocl_kernel_metadata_t;
 
 //############################################################################
 
+#define POCL_MAGIC_1 0xBE8906A1A83D8D23ULL
```
Comment (pjaaskel) [inline lib/CL/pocl_cl.h:167]: What are these? Should be well-documented.

**Review by pjaaskel — CHANGES_REQUESTED** [review]

File: CHANGES [hunk 4]
```diff
@@ -4,6 +4,11 @@
 Notable User Facing Changes
 ---------------------------
 
+- CUDA driver changed to not use spinning in clFinish/clWaitForEvents
+- CMake: OCS_AVAILABLE renamed to ENABLE_LLVM
```
[the remark this answers is not in the document]
Comment (franz) [reply CHANGES:8]: At least the 2nd one is user-facing, though i suspect 99% of builds are with LLVM and only use WITH_LLVM_CONFIG cmake variable.

File: CHANGES [hunk 5]
```diff
@@ -4,6 +4,11 @@
...
+- IPO is enabled on code of runtime library
```
[the remark this answers is not in the document]
Comment (franz) [reply CHANGES:9]: OK

File: CHANGES [hunk 6]
```diff
@@ -4,6 +4,11 @@
...
+- some portability fixes for MSVC / Android
+- LTTng tracing improved
```
[the remark this answers is not in the document]
Comment (franz) [reply CHANGES:11]: It's explained in the commit message. more tracepoints, new tracepoints for synchronous calls. I'll add  it here

File: CMakeLists.txt [hunk 7]
```diff
@@ -1084,25 +1123,20 @@ string(REGEX REPLACE "[\r\n\t ]+" "\", \"" HOST_LD_FLAGS_ARRAY "${HOST_LD_FLAGS_
...
   set(OCL_KERNEL_TARGET "${LLC_TRIPLE}") #The kernel target triplet.
   set(OCL_KERNEL_TARGET_CPU "${LLC_HOST_CPU}") #The kernel target CPU variant.
+  set(BUILD_BASIC 1)
   set(BUILD_PTHREAD 1)
 endif()
 
 # The accel device could be built by default, but it's implemented in C++,
 # thus requires a C++ compiler, so let's not.
 if(ENABLE_ACCEL_DEVICE)
-  set(BUILD_BASIC 1)
```
[the remark this answers is not in the document]
Comment (franz) [reply CMakeLists.txt:1105]: I moved several "shared" driver callbacks to `lib/CL/devices/common_driver.{c,h}`

File: lib/CL/clCreateContext.c [hunk 8]
```diff
@@ -116,8 +114,9 @@ int context_set_properties(cl_context                    context,
     }
 }
 
+extern int pocl_offline_compile;
 unsigned cl_context_count = 0;
-pocl_lock_t pocl_context_handling_lock = POCL_LOCK_INITIALIZER;
```
[the remark this answers is not in the document]
Comment (franz) [reply lib/CL/clCreateContext.c:120]: yes, it's initialized by code in `devices.c`

File: lib/CL/pocl_cl.h [hunk 9]
```diff
@@ -162,12 +175,16 @@ typedef pthread_mutex_t pocl_lock_t;
     POCL_UNLOCK_OBJ (__OBJ__);                  \
   } while (0)
 
+
+extern uint64_t last_object_id;
```
[the remark this answers is not in the document]
Comment (franz) [reply lib/CL/pocl_cl.h:220]: possibly, but that sounds like a thing for another PR

File: lib/CL/pocl_cl.h [hunk 10]
```diff
@@ -659,6 +676,7 @@ typedef enum
 struct _cl_device_id {
   POCL_ICD_OBJECT
   POCL_OBJECT;
+  uint64_t id;
```
[the remark this answers is not in the document]
Comment (franz) [reply lib/CL/pocl_cl.h:834]: good point, will fix.

File: CHANGES (hunk 6 above)
[the remark this answers is not in the document]
Comment (pjaaskel) [reply CHANGES]: I see you added more explanations to the CHANGES, which is what I of course asked for.

File: lib/CL/pocl_util.c (hunk 1 above)
Comment (pjaaskel) [reply lib/CL/pocl_util.c]: I suppose it was removed completely?

File: lib/CL/pocl_cl.h (hunk 3 above)
Comment (pjaaskel) [reply lib/CL/pocl_cl.h:179]: This is still undocumented. Reading the code it's supposed to add extra 'validness' protection by checking if the magic numbers are there or not. Do you find this useful in some case/debugging? It is not nearly fool proof anyhow, like we've discussed in the past. If we allocate a new CL object to that space in another thread, the old reference looks valid even though it was freed in between etc. If you have found it useful, please add comments to the magic numbers or even rename to them to 'sentinel' or similar slightly more descriptive which is a common name for this type trickery.

**Review by pjaaskel — APPROVED** [review]

**Comment by pjaaskel** [comment]
Breaks HSA. https://termbin.com/zsp7p

**Pull request merged**